File: 2015/02.py
# --- Day 2: I Was Told There Would Be No Math ---

paper = ribbon = 0

with open("input/02.txt") as input:
  for line in input:
    dimensions = map(lambda x: int(x), line.split("x")) # "8x5x3" -> [ 8, 5, 3]
    dimensions.sort()                                   # [ 8, 5, 3 ] -> [ 3, 5, 8 ]
    l, w, h = dimensions                                # l = 3, w = 5, h = 8
    paper += (2*l*w + 2*w*h + 2*h*l) + l*w              # (surface) + slack
    ribbon += (2*l+2*w) + l*w*h                         # (wrap) + bow

print paper
print ribbon

File: 2015/02.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
)

func getDimensions(entry string) []int {
	var dimensions []int
	for _, v := range strings.Split(entry, "x") {
		dimension, err := strconv.Atoi(v)
		if err != nil {
			log.Fatal(err)
		}
		dimensions = append(dimensions, dimension)
	}
	sort.Ints(dimensions)
	return dimensions
}

func calculateWrappingPaper(dimensions []int) int {
	return (2 * dimensions[2] * dimensions[0]) + // 2 * l * w
		(2 * dimensions[0] * dimensions[1]) + // 2 * w * h
		(2 * dimensions[1] * dimensions[2]) + // 2 * h * l
		dimensions[0]*dimensions[1] // slack
}

func calculateRibbon(dimensions []int) int {
	return (2*dimensions[0] + 2*dimensions[1]) + // smallest perimeter
		(dimensions[0] * dimensions[1] * dimensions[2]) // volume
}

func main() {
	file, err := os.Open("input/02.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	wrap, ribbon := 0, 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		dimensions := getDimensions(scanner.Text())
		wrap += calculateWrappingPaper(dimensions)
		ribbon += calculateRibbon(dimensions)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println(wrap)
	fmt.Println(ribbon)
}
